lib/roo_on_rails/checks/heroku/app_exists.rb

```diff
@@ -0,0 +1,44 @@
+require 'roo_on_rails/checks/base'
+require 'roo_on_rails/checks/multi'
+
+module RooOnRails
+  module Checks
+    module Heroku
+      class AppExists < Base
+        All = Multi.new(variants: %w[staging production], of: self)
+
+        def initialize(env, **options)
+          super(options)
+          @_env = env
+        end
+        
+        def _intro
+          "Checking if #{bold @_env} app exist..."
+        end
+
+        def _call
+          all_apps = _client.app.list.map { |a| a['name'] }
+          including_name = all_apps.select { |a| a.include?(_state.git_repo) }
+          if including_name.empty?
+            _fail "no apps with names including #{bold _state.git_repo} were detected"
+          end
+
+          correct_app = all_apps.select { |a| a =~ /^(roo-)?#{_state.git_repo}-#{@_env}$/ }
+
+          unless correct_app.one?
+            _hardfail "some apps with name #{bold _state.git_repo} exist, but I can't tell which one is for environment #{bold @_env}"
+          end
+
+          _state.heroku.app![@_env] = correct_app.first
+          _ok "found app #{bold correct_app.first}"
+        end
+
+        private
+
+        def _client
+          _state.heroku.api_client
```

Comment:
`_state` is defined in the base class. It's passed on by the constructor (ie. from the harness); it's a `Hashie::Mash`.

The gist as that (as tentatively explained above): checks can depend on the output of other checks, and they communicate through the `_state` because it'd be very inconvenient to have checks _call_ eachother.

Not the most elegant but it's flexible. [Interactor](https://github.com/collectiveidea/interactor#interactor) does this, for instance. I could call it "context" instead of "state" if that's more clear?